src/celeritas/optical/CerenkovGenerator.hh

```diff
@@ -0,0 +1,158 @@
+    inline CELER_FUNCTION
+    CerenkovGenerator(RefractiveIndex const& refractive_index,
+                      CerenkovDistribution const& dist,
+                      Span<PhotonPrimary> photons);
+
+    // Sample Cerenkov photons from the distribution
+    template<class Generator>
+    inline CELER_FUNCTION void operator()(Generator& rng);
+
+  private:
+    RefractiveIndex const& refractive_index_;
+    CerenkovDistribution const& dist_;
+    Span<PhotonPrimary> photons_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from optical properties and distribution parameters.
+ */
+CELER_FUNCTION
+CerenkovGenerator::CerenkovGenerator(RefractiveIndex const& refractive_index,
+                                     CerenkovDistribution const& dist,
+                                     Span<PhotonPrimary> photons)
+    : refractive_index_(refractive_index), dist_(dist), photons_(photons)
+
+{
+    CELER_EXPECT(refractive_index_.data);
+    CELER_EXPECT(dist_);
+    CELER_EXPECT(photons_.size() == dist_.num_photons);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample Cerenkov photons from the distribution.
+ */
+template<class Generator>
+CELER_FUNCTION void CerenkovGenerator::operator()(Generator& rng)
+{
+    NonuniformGrid<real_type> energy_grid(refractive_index_.data.grid,
+                                          refractive_index_.reals);
+    real_type e_min = energy_grid.front();
+    real_type e_max = energy_grid.back();
+
+    // TODO: rename GenericXsCalculator to something more generic?
+    GenericXsCalculator calc_refractive_index(refractive_index_.data,
+                                              refractive_index_.reals);
+    real_type inv_beta = 2 * constants::c_light
+                         / (dist_.pre.velocity + dist_.post.velocity);
+    real_type cos_max = inv_beta / calc_refractive_index(e_max);
+    real_type sin_max_sq = (1 - cos_max) * (1 + cos_max);
+
+    // Calculate changes over the step
+    Real3 delta_pos = dist_.post.pos;
+    for (int j = 0; j < 3; ++j)
+    {
+        delta_pos[j] -= dist_.pre.pos[j];
+    }
+    real_type delta_num_photons = dist_.post.mean_num_photons
+                                  - dist_.pre.mean_num_photons;
+    real_type delta_velocity = dist_.post.velocity - dist_.pre.velocity;
+
+    for (auto i : range(dist_.num_photons))
+    {
+        // Sample energy and direction
+        real_type energy;
+        real_type cos_theta;
+        UniformRealDistribution<real_type> sample_energy(e_min, e_max);
+        do
+        {
+            energy = sample_energy(rng);
+            cos_theta = inv_beta / calc_refractive_index(energy);
+        } while (!BernoulliDistribution(
+            sin_max_sq / ((1 - cos_theta) * (1 + cos_theta)))(rng));
```

Comment:
I've seen the `(1 - x) * (1 + x)` in a couple places in geant4 code; maybe we should add an algorithm for that (presumably it's to reduce roundoff compared to `1 - x^2`).